# Hypothesis Testing


**Null Hypothesis: $H_0$** : 귀무가설

**Alternative Hypothesis: $H_a$** : 대립가설

test statistic:    
   
$$\frac{Best\ Estimate - Hypothesized\ Estimate}{Standard\ Error\ of\ Estimate}$$ 

In [1]:
import statsmodels.api as sm
import numpy as np
import pandas as pd
import scipy.stats.distributions as dist

## 1) One Population Proportion

### Question :
지난 몇년 간 부모들 중 52%가 전자기기 및 소셜미디어가 십대들의 잠 부족을 야기한다고 믿고있었다. 오늘날, 더 많은 부모들이 전자기기 및 소셜미디어가 십대들의 잠 부족을 야기한다고 믿고 있을까? 

#### 설정

**Population**: Parents with a teenager (age 13-18)  
**Parameter of Interest**: p  
**Null Hypothesis:** p = 0.52  
**Alternative Hypthosis:** p > 0.52 (note that this is a one-sided test)

샘플   
1018 Parents   
56% believe that their teenager’s lack of sleep is caused due to electronics and social media.

In [2]:
n = 1018         # 샘플 수
pnull = .52      # null hypothesis
phat = .56

In [3]:
z_test = (phat- pnull)/np.sqrt((pnull*(1-pnull))/n)
z_test

2.5545334262132955

In [4]:
pvalue = dist.norm.cdf(-np.abs(z_test))
pvalue

0.005316510991822442

#### statsmodel로 계산

In [5]:
sm.stats.proportions_ztest(phat * n, n, pnull, alternative='larger')

(2.571067795759113, 0.005069273865860533)

### 결과

test_stat: 2.57   
p-value: 0.005   
   
p값이 유의 수준 0.05 보다 훨씬 작기 때문에, 귀무가설을 기각할 수 있다.   
대립가설 p > 0.52인 주장을 지지한다.   
오늘날의 부모들은 과거보다 더 전자기기 및 소셜 미디어가 십대들의 잠 부족을 야기한다고 믿고 있다.

# 

## 2) Difference in Population Proportions

### Question :
흑인 부모와 히스패닉 부모를 가진 아이들의 수영 레슨을 받은 비율의 차이가 있을까?

#### 설정

**Populations**: All parents of black children age 6-18 and all parents of Hispanic children age 6-18  
**Parameter of Interest**: p1 - p2, where p1 = black and p2 = hispanic  
**Null Hypothesis:** p1 - p2 = 0  
**Alternative Hypthosis:** p1 - p2 $\neq$ = 0  


샘플   
247명 중 91명(36.8%)의 흑인 부모의 아이들이 수영 레슨을 받았다.   
308명 중 120명(38.9%)의 히스패닉 부모의 아이들이 수영 레슨을 받았다.

In [6]:
# 샘플 수 및 수영 레슨 받은 아이의 수
n1 = 247
n2 = 308
y1 = 91
y2 = 120

In [7]:
# 비율
p1 = round(y1/n1, 2)
p2 = round(y2/n2, 2)
(p1, p2)

(0.37, 0.39)

In [8]:
# combined population proportion
phat = (y1+y2)/(n1+n2)
phat

0.3801801801801802

In [9]:
# variance of the combined population proportion
va = phat * (1 - phat)
va

0.2356432107783459

In [10]:
# standard error of the combined population proportion
se = np.sqrt(va * (1 / n1 + 1 / n2))
se

0.04146198808473324

In [11]:
# Test statistic and its p-value
test_stat = (p1 - p2) / se
pvalue = 2*dist.norm.cdf(-np.abs(test_stat))
(test_stat, pvalue)

(-0.4823695371077548, 0.6295434573871281)

#### statsmodel로 계산

In [12]:
count = np.array([y1, y2])
nobs = np.array([n1, n2])
sm.stats.proportions_ztest(count,nobs, alternative='two-sided')

(-0.5110545335044571, 0.6093128715165157)

### 결과

test_stat: -0.48   
p-value: 0.63   
   
p값이 유의 수준 0.05 보다 훨씬 크기 때문에, 귀무가설을 기각할 수 없다.   
p1 - p2가 0이 아니라는 증거를 찾을 수 없다.   
p1 - p2 = 0 이라는 귀무가설을 지지한다.   
흑인과 히스패닉 아이들의 수영 레슨을 받은 비율은 차이가 없다.


# 

## 3) One Population Mean

### Question :
어른들의 수레바퀴(측면 회전 운동, 옆구르기) 평균 거리는 80인치 보다 클까?

#### 설정

**Population**: All adults  
**Parameter of Interest**: $\mu$, population mean cartwheel distance.   
**Null Hypothesis:** $\mu$ = 80   
**Alternative Hypthosis:** $\mu$ > 80

25 Adults

$\mu = 82.46$

$\sigma = 15.06$

In [13]:
df = pd.read_csv("Cartwheeldata.csv")
df

ID  Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0    1   56      F            1       Y             1   62.00      61.0   
1    2   26      F            1       Y             1   62.00      60.0   
2    3   33      F            1       Y             1   66.00      64.0   
3    4   39      F            1       N             0   64.00      63.0   
4    5   27      M            2       N             0   73.00      75.0   
5    6   24      M            2       N             0   75.00      71.0   
6    7   28      M            2       N             0   75.00      76.0   
7    8   22      F            1       N             0   65.00      62.0   
8    9   29      M            2       Y             1   74.00      73.0   
9   10   33      F            1       Y             1   63.00      60.0   
10  11   30      M            2       Y             1   69.50      66.0   
11  12   28      F            1       Y             1   62.75      58.0   
12  13   25      F            1       Y             1   65.00      64.5   
13  14   23      F            1       N             0   61.50      57.5   
14  15   31      M            2       Y             1   73.00      74.0   
15  16   26      M            2       Y             1   71.00      72.0   
16  17   26      F            1       N             0   61.50      59.5   
17  18   27      M            2       N             0   66.00      66.0   
18  19   23      M            2       Y             1   70.00      69.0   
19  20   24      F            1       Y             1   68.00      66.0   
20  21   23      M            2       Y             1   69.00      67.0   
21  22   29      M            2       N             0   71.00      70.0   
22  23   25      M            2       N             0   70.00      68.0   
23  24   26      M            2       N             0   69.00      71.0   
24  25   23      F            1       Y             1   65.00      63.0   

    CWDistance Complete  CompleteGroup  Score  
0           79        Y              1      7  
1           70        Y              1      8  
2           85        Y              1      7  
3           87        Y              1     10  
4           72        N              0      4  
5           81        N              0      3  
6          107        Y              1     10  
7           98        Y              1      9  
8          106        N              0      5  
9           65        Y              1      8  
10          96        Y              1      6  
11          79        Y              1     10  
12          92        Y              1      6  
13          66        Y              1      4  
14          72        Y              1      9  
15         115        Y              1      6  
16          90        N              0     10  
17          74        Y              1      5  
18          64        Y              1      3  
19          85        Y              1      8  
20          66        N              0      2  
21         101        Y              1      8  
22          82        Y              1      4  
23          63        Y              1      5  
24          67        N              0      3

In [14]:
n = len(df)                        # 샘플 수
mean = df["CWDistance"].mean()     # 평균
sd = df["CWDistance"].std()        # 표준편차
(n, mean, sd)

(25, 82.48, 15.058552387264855)

In [15]:
# standard error
se = sd/np.sqrt(n)

In [16]:
# one sample t-test
t_test = (mean-80)/se
t_test

0.8234523266982027

In [17]:
pvalue = dist.norm.cdf(-np.abs(t_test))
pvalue

0.20512540845395272

#### statsmodel로 계산

In [18]:
sm.stats.ztest(df["CWDistance"], value = 80, alternative = "larger")

(0.8234523266982029, 0.20512540845395266)

### 결과

test_stat: 0.82   
pvalue: 0.21   
   
p값이 유의 수준 0.05 보다 훨씬 크기 때문에, 귀무가설을 기각할 수 없다.   
평균 거리가 80이 아니라는 증거를 찾을 수 없다.   
평균 거리 = 80 이라는 귀무가설을 지지한다.   
어른들의 옆 구르기의 평균 거리는 80인치로 추정된다.

# 

## 4) Difference in Population Means

### Question :
NHANES 데이터에서, 남성들의 Body Mass Index(BMI) 수치가 여성보다 더 클까?   
   


#### 설정
**Population**: Adults in the NHANES data.  
**Parameter of Interest**: $\mu_1 - \mu_2$, Body Mass Index.  
**Null Hypothesis:** $\mu_1 = \mu_2$  
**Alternative Hypthosis:** $\mu_1 \neq \mu_2$


In [19]:
url = "nhanes_2015_2016.csv"
da = pd.read_csv(url)
da.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

In [20]:
females = da[da["RIAGENDR"] == 2]
male = da[da["RIAGENDR"] == 1]

In [21]:
# 여성
n1 = len(females)
mu1 = females["BMXBMI"].mean()
sd1 = females["BMXBMI"].std()

(n1, mu1, sd1)

(2976, 29.939945652173996, 7.75331880954568)

In [22]:
# 남성
n2 = len(male)
mu2 = male["BMXBMI"].mean()
sd2 = male["BMXBMI"].std()

(n2, mu2, sd2)

(2759, 28.778072111846985, 6.252567616801485)

In [23]:
se = np.sqrt(sd1**2/n1 + sd2**2/n2)
se

0.18538992862064485

In [24]:
test_stat = (mu1-mu2)/se
test_stat

6.267188023490217

In [25]:
pvalue = 2*dist.norm.cdf(-np.abs(test_stat))
pvalue

3.676261615853684e-10

#### statsmodel로 계산

In [26]:
sm.stats.ztest(females["BMXBMI"].dropna(), male["BMXBMI"].dropna())

(6.1755933531383205, 6.591544431126401e-10)

### 결과

test_stat: 6.18  
pvalue: 6.591544431126401e-10   
   
p값이 거의 0에 가깝게 작다.   
귀무가설을 기각할 수 있다. 대립가설을 지지한다.   
남성의 평균 BMI와 여성의 평균 BMI는 같다는 것은 매우 일반적이지 않다.   
남성의 평균 BMI와 여성의 평균 BMI는 다를 것으로 추정된다.